nemo_aligner/models/mm/stable_diffusion/megatron_sd_draftp_model.py

```diff
@@ -0,0 +1,381 @@
+        self.model.with_distributed_adam = False
+
+    def finish_inference(self):
+        return
+
+    def finish_training_step(self):
+        grad_reductions(self)
+
+    def infer(self):
+        return
+
+    def prepare_for_inference(self):
+        return
+
+    def prepare_for_training_step(self):
+        # custom trainers will always zero grad for us
+        prepare_for_training_step(self.model, zero_grad=False)
+
+    @torch.no_grad()
+    def generate_log_images(self, latents, batch, model):
+
+        with torch.cuda.amp.autocast(
+            enabled=self.autocast_dtype in (torch.half, torch.bfloat16), dtype=self.autocast_dtype,
+        ):
+
+            sampler = sampling_utils.initialize_sampler(model, self.sampler_type.upper())
+
+            batch_size = len(batch)
+            cond, u_cond = sampling_utils.encode_prompt(
+                model.cond_stage_model, batch, self.unconditional_guidance_scale
+            )
+
+            _, intermediates = sampler.sample(
+                S=self.inference_steps,
+                conditioning=cond,
+                batch_size=batch_size,
+                shape=list(latents.shape[:1] + latents.shape[2:]),
+                verbose=False,
+                unconditional_guidance_scale=self.unconditional_guidance_scale,
+                unconditional_conditioning=u_cond,
+                eta=self.eta,
+                x_T=latents,
+                log_every_t=1,
+                truncation_steps=0,
+                return_logprobs=False,
+                return_mean_var=False,
+            )
+
+            # # transforms trajectores from a list of T+1 (b x image) tensors to a tensor of shape (b * (T+1), ...)
+            trajectories_predx0 = (
+                torch.stack(intermediates["pred_x0"], dim=0)
+                .transpose(0, 1)
+                .contiguous()
+                .view(-1, *intermediates["pred_x0"][0].shape[1:])
+            )
+
+            vae_decoder_output = []
+            idx_denoised_imgs = [self.inference_steps + (self.inference_steps + 1) * i for i in range(batch_size)]
+
+            for i in range(0, batch_size, self.vae_batch_size):
+                image = self.model.differentiable_decode_first_stage(
+                    trajectories_predx0[idx_denoised_imgs[i : i + self.vae_batch_size]]
+                )
+
+                vae_decoder_output.append(image)
+
+            vae_decoder_output = torch.cat(vae_decoder_output, dim=0)
+            vae_decoder_output = torch.clip((vae_decoder_output + 1) / 2, 0, 1) * 255.0
+
+            log_reward = [
+                self.reward_model.get_reward(
+                    vae_decoder_output[i].unsqueeze(0).detach().permute(0, 2, 3, 1), batch  # (C, H, W) -> (1, H, W, C)
+                ).item()
+                for i in range(batch_size)
+            ]
+            log_img = [
+                np.transpose(vae_decoder_output[i].float().detach().cpu().numpy(), (1, 2, 0))  # (C, H, W) -> (H, W, C)
```

Comment:
nit: why not do the transpose in torch and then convert it to numpy?

Reply:
done